tidb/features.go

```diff
@@ -0,0 +1,46 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tidb
+
+const (
+	// FeatureIDTiDB represents the general TiDB-specific features.
+	FeatureIDTiDB = ""
+	// FeatureIDAutoRandom is the `auto_random` feature.
+	FeatureIDAutoRandom = "auto_rand"
+	// FeatureIDAutoIDCache is the `auto_id_cache` feature.
+	FeatureIDAutoIDCache = "auto_id_cache"
+	// FeatureIDAutoRandomBase is the `auto_random_base` feature.
+	FeatureIDAutoRandomBase = "auto_rand_base"
+	// FeatureIDClusteredIndex is the `clustered_index` feature.
+	FeatureIDClusteredIndex = "clustered_index"
+	// FeatureIDForceAutoInc is the `force auto_increment` feature.
+	FeatureIDForceAutoInc = "force_inc"
+)
+
+var featureIDs = map[string]struct{}{
+	FeatureIDAutoRandom:     {},
+	FeatureIDAutoIDCache:    {},
+	FeatureIDAutoRandomBase: {},
+	FeatureIDClusteredIndex: {},
+	FeatureIDForceAutoInc:   {},
+}
+
+func CanParseFeature(fs ...string) bool {
```

Comment:
The old code is more general, the parser is not specific to TiDB and you can register the feature to control it's behavior ...

Reply:
OK, since you are the original author of this code, if you thinks it's OK, I have no problem with it.

Reply:
> the parser is not specific to TiDB

But the keywords(`AUTO_RANDOM`, `CLUSTERED_INDEX`...) are TiDB-specific. It is reasonable to attach these keywords to fixed feature IDs.

> you can register the feature to control it's behavior

The register mechanism is not that convenient. If the user registers a new feature ID(`/*T![new_feature_id] */`), he must also modify the code in the parser to add a new keyword to wrap. If the user doesn't need the new keyword, he should use `/*T xxx */` instead.